OrbitVulkanLayer/SubmissionTracker.h

```diff
@@ -0,0 +1,779 @@
+  // This allows us to map submissions from the Vulkan layer to the driver submissions.
+  [[nodiscard]] std::optional<internal::QueueSubmission> PersistCommandBuffersOnSubmit(
+      uint32_t submit_count, const VkSubmitInfo* submits) {
+    if (!IsCapturing()) {
+      // `PersistDebugMarkersOnSubmit` and `OnCaptureFinished` will take care of clean up and
+      // slot resetting.
+      return std::nullopt;
+    }
+
+    internal::QueueSubmission queue_submission;
+    queue_submission.meta_information.pre_submission_cpu_timestamp = MonotonicTimestampNs();
+    queue_submission.meta_information.thread_id = GetCurrentThreadId();
+
+    for (uint32_t submit_index = 0; submit_index < submit_count; ++submit_index) {
+      VkSubmitInfo submit_info = submits[submit_index];
+      queue_submission.submit_infos.emplace_back();
+      internal::SubmitInfo& submitted_submit_info = queue_submission.submit_infos.back();
+      for (uint32_t command_buffer_index = 0; command_buffer_index < submit_info.commandBufferCount;
+           ++command_buffer_index) {
+        VkCommandBuffer command_buffer = submit_info.pCommandBuffers[command_buffer_index];
+        CHECK(command_buffer_to_state_.contains(command_buffer));
+        CommandBufferState& state = command_buffer_to_state_.at(command_buffer);
+
+        // If we neither recorded the end nor the begin of a command buffer, we have no information
+        // to send.
+        if (!state.command_buffer_end_slot_index.has_value()) {
+          CHECK(!state.command_buffer_end_slot_index.has_value());
+          continue;
+        }
+
+        internal::SubmittedCommandBuffer submitted_command_buffer{
+            .command_buffer_begin_slot_index = state.command_buffer_begin_slot_index,
+            .command_buffer_end_slot_index = state.command_buffer_end_slot_index.value()};
+        submitted_submit_info.command_buffers.emplace_back(submitted_command_buffer);
+
+        // Remove the slots from the state now, such that `OnCaptureFinished` wont reset the slots
+        // twice. Note, that they will be reset in `CompleteSubmits`.
+        state.command_buffer_begin_slot_index.reset();
+        state.command_buffer_end_slot_index.reset();
+      }
+    }
+
+    return queue_submission;
+  }
+
+  // We assume to be capturing if `queue_submission_optional` has content, i.e. we were capturing
+  // on this VkQueueSubmit before calling into the driver.
+  // It also takes a timestamp after the execution of the driver code for the submission.
+  // This allows us to map submissions from the Vulkan layer to the driver submissions.
+  void PersistDebugMarkersOnSubmit(
+      VkQueue queue, uint32_t submit_count, const VkSubmitInfo* submits,
+      std::optional<internal::QueueSubmission> queue_submission_optional) {
+    absl::WriterMutexLock lock(&mutex_);
+    if (!queue_to_markers_.contains(queue)) {
+      queue_to_markers_[queue] = {};
+    }
+    CHECK(queue_to_markers_.contains(queue));
```

Comment:
Isn't this a bit redundant? It was just added.

Reply:
I decided to keep this check, while writing this code. If the code gets modified to e.g. add the queue_to_markers on a different place, we still have that check here. In fact, for development those CHECKs are much nicer for debugging than the exception when when calling the `.at` if not present. As this is only once every submission, I'd just keep that one.